frame/assets/src/lib.rs

```diff
@@ -247,19 +256,594 @@ decl_error! {
+		///
+		/// Weight: `O(z)` where `z` is the number of zombie accounts.
+		#[weight = T::WeightInfo::destroy(*zombies_witness)]
+		fn destroy(origin,
+			#[compact] id: T::AssetId,
+			#[compact] zombies_witness: u32,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			Asset::<T>::try_mutate_exists(id, |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::Unknown)?;
+				ensure!(details.owner == origin, Error::<T>::NoPermission);
+				ensure!(details.accounts == details.zombies, Error::<T>::RefsLeft);
+				ensure!(details.zombies <= zombies_witness, Error::<T>::BadWitness);
+				T::Currency::unreserve(&details.owner, details.deposit);
+
+				*maybe_details = None;
+				Account::<T>::remove_prefix(&id);
+				Self::deposit_event(RawEvent::Destroyed(id));
+				Ok(())
+			})
+		}
+
+		/// Destroy a class of fungible assets.
+		///
+		/// The origin must conform to `ForceOrigin`.
+		///
+		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
+		/// asset.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[weight = T::WeightInfo::force_destroy(*zombies_witness)]
+		fn force_destroy(origin,
+			#[compact] id: T::AssetId,
+			#[compact] zombies_witness: u32,
+		) -> DispatchResult {
+			T::ForceOrigin::ensure_origin(origin)?;
+
+			Asset::<T>::try_mutate_exists(id, |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::Unknown)?;
+				ensure!(details.accounts == details.zombies, Error::<T>::RefsLeft);
+				ensure!(details.zombies <= zombies_witness, Error::<T>::BadWitness);
+				T::Currency::unreserve(&details.owner, details.deposit);
+
+				*maybe_details = None;
+				Account::<T>::remove_prefix(&id);
+				Self::deposit_event(RawEvent::Destroyed(id));
+				Ok(())
+			})
+		}
+
+		/// Mint assets of a particular class.
+		///
+		/// The origin must be Signed and the sender must be the Issuer of the asset `id`.
+		///
+		/// - `id`: The identifier of the asset to have some amount minted.
+		/// - `beneficiary`: The account to be credited with the minted assets.
+		/// - `amount`: The amount of the asset to be minted.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(1)`
+		/// Modes: Pre-existence of `beneficiary`
+		#[weight = T::WeightInfo::mint()]
+		fn mint(origin,
+			#[compact] id: T::AssetId,
+			beneficiary: <T::Lookup as StaticLookup>::Source,
+			#[compact] amount: T::Balance
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			Asset::<T>::try_mutate(id, |maybe_details| {
+				let d = maybe_details.as_mut().ok_or(Error::<T>::Unknown)?;
+
+				ensure!(&origin == &d.issuer, Error::<T>::NoPermission);
+				d.supply = d.supply.saturating_add(amount);
+
+				Account::<T>::try_mutate(id, &beneficiary, |t| -> DispatchResult {
+					let new_balance = t.balance.saturating_add(amount);
```

Comment:
Should both be checked_add right?

Reply:
Here is a test which illustrates the problem:

```
	#[test]
	fn assets_handles_overflows_well() {
		new_test_ext().execute_with(|| {
			assert_ok!(Assets::force_create(Origin::root(), 0, 1, 10, 1));
			assert_ok!(Assets::mint(Origin::signed(1), 0, 1, u64::max_value()));
			assert_eq!(Assets::balance(0, 1), u64::max_value());
			// This is bad!
			assert_ok!(Assets::mint(Origin::signed(1), 0, 2, u64::max_value()));
			assert_eq!(Assets::balance(0, 2), u64::max_value());
			assert_eq!(Assets::total_supply(0), u64::max_value());
			assert_ok!(Assets::burn(Origin::signed(1), 0, 2, u64::max_value()));
			assert_eq!(Assets::total_supply(0), 0);
			// This is really bad!  ^    v
			assert_eq!(Assets::balance(0, 1), u64::max_value());
		});
	}
```

The second mint shouldnt pass imo. If we use `checked_add` when always increasing the balances, and `saturating_sub` when decreasing, I think we will get a better overall behavior

Reply:
Yeah, supply is better as checked add, though given it's the same datatype, there's no reason to use checked add on the account's balance.